javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -175,10 +92,11 @@ public M distinctBy(Comparator<? super Tuple2<K, V>> comparator) {
         return (M) createFromEntries(iterator().distinctBy(comparator));
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public <U> Multimap<K, V> distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
+    public <U> M distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
         Objects.requireNonNull(keyExtractor, "keyExtractor is null");
-        return createFromEntries(iterator().distinctBy(keyExtractor));
+        return (M) createFromEntries(iterator().distinctBy(keyExtractor));
```

Comment:
The class AbstractMultimap is not relevant any more. You can ignore it. It is just a Clipboard for the ongoing work.

E.g. this method will be moved to `Multimaps` and get additional parameters, one of them is a function `OfEntries` that is type-safe. The cast will disappear.